### PR TITLE
Nuclear Emergency: Disky's Revenge

### PR DESCRIPTION
**Pull request by francinum**
## About The Pull Request

Adds Nuclear Defense, an alternative version of Nuclear Emergency.

The nuclear bomb in this mode:
* Does not require the disk, but does recieve a substantial speed boost from it.
* Has a fixed timer of 10 minutes, with the disk serving to speed this up
* Can be broken. Currently has a DT of 10, and a health of 2000.
* **Must be planted in one of 3 randomly selected areas.**
* **Cannot be disarmed once planted.**

## Cleanroom Design Disclaimer

The changes made within this PR were made without reference to goon's code, though I have viewed it in the past, their system is of little help to ours.

The code in this commit is wholly my own, and is ~~with regret~~ licensed under the AGPL.

## TODO

- [ ] Fix the round not ending on nuke destruction
- [ ] Make sure round ends on nuke explosion, Surprisingly didn't test that.
- [ ] Maybe write a new tips pane, Not sure how to show it though...
- [ ] More testing and balance passing on the tweakables.

## Changelog
:cl:
add: Nuclear Defense, an alternative flavor of Nuclear Emergency.
/:cl:

### CODE REVIEW

File: code/game/gamemodes/nuclear/nuclear_defense.dm [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+/datum/team/nuclear/defense/get_result()
+	//This is essentially a remapper, as NOSURVIVORS is the intended victory condition.
+	. = ..()
+	switch(.)
+		if(NUKE_RESULT_NOSURVIVORS)
+			return NUKE_RESULT_NUKE_WIN
+		if(NUKE_RESULT_DISK_LOST)
+			return NUKE_RESULT_CREW_WIN
+		else
+			return //Everything else can fall through
```
Comment (utilbird) [inline code/game/gamemodes/nuclear/nuclear_defense.dm:47]:
```suggestion
	return //Everything else can fall through
```

File: code/modules/antagonists/nukeop/equipment/nuclearbomb.dm [hunk 2]
```diff
@@ -239,6 +239,8 @@
 			playsound(loc, 'sound/items/timer.ogg', volume, FALSE)
 
 /obj/machinery/nuclearbomb/proc/update_ui_mode()
+	if(!src)
+		return
```
Comment (utilbird) [inline code/modules/antagonists/nukeop/equipment/nuclearbomb.dm:243]: I _believe_ this is redundant, especially at the start of a proc

File: code/modules/antagonists/nukeop/equipment/nuclearbomb.dm [hunk 3]
```diff
@@ -603,6 +607,206 @@ This is here to make the tiles around the station mininuke change when it's arme
...
+
+/obj/machinery/nuclearbomb/syndicate/proto/ui_interact(mob/user, datum/tgui/ui=null)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "GoonNuke")
+		ui.open()
+
+/obj/machinery/nuclearbomb/syndicate/proto/update_ui_mode()
+	if(!src)
+		return
```
Comment (utilbird) [inline code/modules/antagonists/nukeop/equipment/nuclearbomb.dm:652]: also might be redundant

File: code/modules/antagonists/nukeop/equipment/nuclearbomb.dm [hunk 4]
```diff
@@ -603,6 +607,206 @@ This is here to make the tiles around the station mininuke change when it's arme
...
+	if(timing)
+		priority_announce("A high-yield fusion warhead has been detected in [get_area(src).name]. The device must be destroyed before the warhead detonates!","Nuclear Emergency", 'sound/misc/airraid.ogg')
+
+
+/obj/machinery/nuclearbomb/syndicate/proto/disk_check(obj/item/disk/nuclear/D)
+	if(D.fake)
+		say("Coprocessor CPUID Invalid; disk not recognised.")
+		return FALSE
+	else
+		return TRUE
```
Comment (utilbird) [inline code/modules/antagonists/nukeop/equipment/nuclearbomb.dm:751]:
```suggestion
	return TRUE
```

Comment (utilbird) [inline code/modules/antagonists/nukeop/equipment/nuclearbomb.dm:751]: This is almost identical to the parent proc besides the say message which is only _slightly_ different. Not that big of a problem, but a nitpick

File: code/modules/antagonists/nukeop/equipment/nuclearbomb.dm [hunk 5]
```diff
@@ -603,6 +607,206 @@ This is here to make the tiles around the station mininuke change when it's arme
...
+
+/obj/machinery/nuclearbomb/syndicate/proto/ui_act(action, params)
+	if(..()) //I just want to call the grandparent here...
+		return
+	playsound(src, "terminal_type", 20, FALSE)
+	switch(action)
+		if("eject_disk")
+			if(auth && auth.loc == src)
```
Comment (utilbird) [inline code/modules/antagonists/nukeop/equipment/nuclearbomb.dm:759]:
```suggestion
			if(auth?.loc == src)
```

File: code/modules/antagonists/nukeop/equipment/nuclearbomb.dm [hunk 6]
```diff
@@ -603,6 +607,206 @@ This is here to make the tiles around the station mininuke change when it's arme
...
+	else
+		return NUKE_OFF_LOCKED
+
+/obj/machinery/nuclearbomb/syndicate/proto/process()
+	if(timing && !exploding)
+		if(auth) //Do we have the disk/coprocessor?
+			detonation_timer -= disky_speedup //Subtract the disky bonus
+		var/remaining = get_time_left()
+		if(remaining > 0)
+			remaining = "[add_leading(num2text((remaining / 60) % 60), 2, "0")]:[add_leading(num2text(remaining % 60), 2, " ")]"
```
Comment (utilbird) [inline code/modules/antagonists/nukeop/equipment/nuclearbomb.dm:805]: Surprised no one's made this a define yet given all it's occurrences 

_(wink wink)_

File: code/modules/antagonists/nukeop/equipment/nuclearbomb.dm [hunk 7]
```diff
@@ -603,6 +607,206 @@ This is here to make the tiles around the station mininuke change when it's arme
...
+
+/obj/machinery/nuclearbomb/syndicate/proto/get_nuke_state()
+	if(exploding)
+		return NUKE_ON_EXPLODING
+	if(timing)
+		return NUKE_ON_TIMING
+	if(ui_mode == PROTONUKEUI_NOT_READY)
+		return NUKE_OFF_UNLOCKED
+	else
+		return NUKE_OFF_LOCKED
```
Comment (utilbird) [inline code/modules/antagonists/nukeop/equipment/nuclearbomb.dm:797]:
```suggestion
	return NUKE_OFF_LOCKED
```

File: code/modules/antagonists/nukeop/equipment/nuclearbomb.dm [hunk 8]
```diff
@@ -603,6 +607,206 @@ This is here to make the tiles around the station mininuke change when it's arme
...
+				return
+			else
+				ui_mode = PROTONUKEUI_NOT_READY
+				playsound(src, 'sound/machines/nuke/confirm_beep.ogg', 50, FALSE)
+				update_icon()
+				return
+	else
+		ui_mode = PROTONUKEUI_WRONGZONE
+		update_icon()
+		return
```
Comment (utilbird) [inline code/modules/antagonists/nukeop/equipment/nuclearbomb.dm:686]: Most of these return statements aren't needed

**Comment by Cartlord** [comment]
What will be the pop requirement?

**Comment by eeSPee** [comment]
IMO this should be a uplink item for nukies and a option instead of a different game mode.

**Comment by francinum** [comment]
> IMO this should be a uplink item for nukies and a option instead of a different game mode.

For code reasons related to the way the gamemode code needs to work, this isn't an acceptable option.

**Comment by eeSPee** [comment]
Fair enough.

**Comment by DatBoiTim** [comment]
So is this meant to be a suicide mission or is it like classic nukies where you are expected to survive the blast as nukie?

Also Wiki Updates

**Comment by francinum** [comment]
> So is this meant to be a suicide mission or is it like classic nukies where you are expected to survive the blast as nukie?
> 
> Also Wiki Updates

Oh you're expected to die.

**Comment by DatBoiTim** [comment]
Also is it forced war ops or is stealth a thing? Stealth could be really easy since there's no disk to acquire

**Comment by ghost** [comment]
`The changes made within this PR were made without reference to goon's code,`

the quality reminds me of goon though

**Comment by heepox** [comment]
ah yes the goon variety of Nukies.

**Comment by FortSumter** [comment]
I kinda would like this to be the default way nuke ops works instead of a new game mode. The current way is sort of stupid that we need to get the captain's disk to activate our own nuke, and are expected to escape with our lives.

**Comment by ike709** [comment]
Stale.

**Pull request closed without merging**